File: solutions/cpp/atbash-cipher/1/atbash_cipher.cpp
#include "atbash_cipher.h"

#include <cctype>

namespace atbash_cipher
{
    std::string encode(const std::string &plaintext)
    {
        std::string ciphertext = "";

        for (char c : plaintext)
        {
            if (!std::isalnum(c))
                continue;

            size_t ciphertext_length = ciphertext.length();
            if (ciphertext_length > 0 && (ciphertext_length + 1) % 6 == 0)
            {
                ciphertext.push_back(' ');
            }

            if (std::isalpha(c))
            {
                ciphertext.push_back('a' + 'z' - std::tolower(c));
            }
            else
            {
                ciphertext.push_back(c);
            }
        }
        return ciphertext;
    }

    std::string decode(const std::string &ciphertext)
    {
        std::string plaintext = "";

        for (char c : ciphertext)
        {
            if (!std::isalnum(c))
                continue;

            if (std::isalpha(c))
            {
                plaintext.push_back('a' + ('z' - std::tolower(c)));
            }
            else
            {
                plaintext.push_back(c);
            }
        }
        return plaintext;
    }

} // namespace atbash_cipher

File: solutions/python/atbash-cipher/1/atbash_cipher.py
def _atbash_char(char: str) -> str:
    if char.isalpha():
        return chr(ord('z') - (ord(char.lower()) - ord('a')))
    elif char.isdigit():
        return char
    else:
        return ""

def _atbash_str(text: str) -> str:
    return "".join(_atbash_char(char) for char in text if char.isalnum())

def encode(plain_text: str) -> str:
    result = _atbash_str(plain_text)
    return " ".join([result[i:i+5] for i in range(0, len(result), 5)])

def decode(ciphered_text: str) -> str:
    return _atbash_str(ciphered_text)
